apps/src/lab2/levelEditors/aichatSettings/FieldSection.tsx

```diff
@@ -47,7 +47,14 @@ const FieldSection: React.FunctionComponent<FieldSectionProps> = ({
               id={fieldName}
               type={inputType}
               value={initialCustomizations[fieldName] as string | number}
-              onChange={e => setPropertyValue(fieldName, e.target.value)}
+              onChange={e =>
+                setPropertyValue(
+                  fieldName,
+                  fieldName === 'temperature'
+                    ? parseFloat(e.target.value)
+                    : e.target.value
+                )
+              }
```

Comment:
This is a dangerous pattern, since we're relying upon the calling function to properly sanitize the data _and_ we're doing it in at least two different places in at least two different ways. (as part of `setAiCustomizationProperty` _and_ `setPropertyValue`.). I'm worried that these two things can get out of sync - what if one is updated and another isn't?

I'd recommend a pattern where the data is sanitized when it comes in. Ideally, I'd prefer an explicit `setTemperature` setter that's used universally and only have _that_ sanitize its own inputs, but we're a few degrees abstracted from that.

Instead, we can create a centralized `sanitizeValue` method. Something like this:

```
const sanitizeValue = (property, value) => {
  if (property === 'temperature') {
    return parseFloat(e.target.value);
  }
  else {
    return value;
  }
}
```

This is kinda messy in that that `sanitizeValue` function will end up having centralized knowledge of all properties (eventually, assuming all inputs need to be sanitized in different ways).

But then, internally, `setPropertyValue` and `setAiCustomizationProperty` can both do:
```
const sanitizedValue = sanitizeValue(property, value)
// do whatever with sanitizedValue
```

and then we don't need to explicitly do the casting externally and hopefully remember.

---

Alternatively, maybe there's a way the typing could be adapted to allow this? Some weird slicing and dicing to enforce that the property/value must match the defined types? I think it's beyond my typescript abilities atm. I'm not sure how complicated the typing would end up being. But even if the typing is a mess, this would overall be the cleanest approach.

We _could_ overload both functions to enforce the type safety with every case, but since we have two functions involved we're back to the same situation where we've got duplicated values.

Anyway, again, it's sorta fine as is in that it works, but we're introducing risk and I think we should have wider discussiona bout it.

Reply:
Wait, wait, I think I got it! Here's a simplified version of the typing we'd need:

```
// this is the `interface AiCustomizations` from apps/src/aichat/types.ts, simplified for sake of example
interface AiCustomizations {
  temperature: number;
  systemPrompt: string;
}

Right now, both `setPropertyValue` and the action of `setAiCustomizationProperty` are typed as:

  property: keyof AiCustomizations,
  value: AiCustomizations[keyof AiCustomizations]

Instead, re-type as:
  property: T,
  value: AiCustomizations[T]

And add on the generic to the function definition, e.g.:

<T extends keyf AiCustomizations>
```

The generic used for `property` will have to be a key of `AiCustomizations`, _and_ the value will have to be whatever the type of _that specific key_ is.

With this typing in place, _then_ we'd be safe doing the `parseFloat()` redundantly when the function is called, because we're ensuring that the type has to match it, and if a `parseFloat()` is ever missed in a call for `temperature`, we'll get a typescript error instead.

In turn, that prevents us from needing the centralized `sanitizeValue` function at all. 🎉

Reply:
Thanks Jim! Cool that we can type in this way to ensure value matches field type for these two different setter functions. Updated!

Reply:
Heck yeah generics 😎